db/db.go

```diff
@@ -245,9 +239,289 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
 	return newData, newDataChecksum, output, isWrite, nil
 }
 
+func removeTransaction(tx ReadTransaction) {
+	transactionMutex.Lock()
+	defer transactionMutex.Unlock()
+	delete(transactions, tx.ID())
+}
+
+func (db *DB) BeginReadTransaction() ReadTransaction {
+	transactionMutex.Lock()
+	defer transactionMutex.Unlock()
+	txID := transactionCounter
+	transactionCounter++
+	tx := &readTransaction{txID, db.noms, db.head}
+	transactions[txID] = tx
+	return tx
+}
+
+func (db *DB) ReadTransaction(body func(tx ReadTransaction)) error {
+	tx := db.BeginReadTransaction()
+	body(tx)
+	return tx.Close()
+}
+
+func (db *DB) CloseTransaction(txID int) error {
+	tx, err := db.FindReadTransaction(txID)
+	if err != nil {
+		return err
+	}
+	return tx.Close()
+}
+
+func (db *DB) CommitTransaction(txID int) (types.Ref, error) {
+	tx, err := db.FindWriteTransaction(txID)
+	if err != nil {
+		return types.Ref{}, err
+	}
+	return tx.Commit()
+}
+
+func (db *DB) AbortTransaction(txID int) error {
+	tx, err := db.FindWriteTransaction(txID)
+	if err != nil {
+		return err
+	}
+	return tx.Abort()
+}
+
+func (db *DB) BeginWriteTransaction() WriteTransaction {
+	transactionMutex.Lock()
+	defer transactionMutex.Unlock()
+	txID := transactionCounter
+	transactionCounter++
+	tx := &writeTransaction{txID, db, db.noms, db.head, db.head.Data(db.noms).Edit(), false}
+	transactions[txID] = tx
+	return tx
+}
+
+func (db *DB) WriteTransaction(body func(tx WriteTransaction)) error {
```

Comment:
I think I would pull this and the corresponding `ReadTransaction` out of the `db` package just to make the API a little cleaner and easier to grok. Currently the only callers are test code. In that case just put a utility in a test file.

Reply:
Removed alltogether